library/alloc/src/alloc.rs

```diff
@@ -342,3 +362,30 @@ pub fn handle_alloc_error(layout: Layout) -> ! {
     }
     unsafe { oom_impl(layout) }
 }
+
+#[cfg(not(any(test, bootstrap)))]
+#[doc(hidden)]
+#[allow(unused_attributes)]
+#[unstable(feature = "alloc_internals", issue = "none")]
+pub mod __default_lib_allocator {
+    use crate::alloc::Layout;
+
+    // called via generated `__rust_alloc_error_handler`
+
+    // if there is no `#[alloc_error_handler]`
+    #[rustc_std_internal_symbol]
+    pub unsafe extern "C" fn __rdl_oom(size: usize, _align: usize) -> ! {
+        panic!("memory allocation of {} bytes failed", size)
+    }
+
+    // if there is a `#[alloc_error_handler]`
+    #[rustc_std_internal_symbol]
+    pub unsafe extern "C" fn __rg_oom(size: usize, align: usize) -> ! {
```

Comment:
Nothing like this exists for the allocator, right? There the generated `__rust_alloc` etc symbols directly call the lang item? Looks like this is needed because the argument types differ; is there any plan to get rid of that mismatch?

Reply:
Ah no, the `__rd` functions are just generated by the [macro expansion code](https://github.com/rust-lang/rust/blob/85fbf49ce0e2274d0acf798f6e703747674feec3/compiler/rustc_builtin_macros/src/global_allocator.rs#L75).

Reply:
This can be removed, if `compiler/rustc_codegen_llvm/src/allocator.rs` calls directly into the lang item, but with the `extern "Rust"` API and a `Layout` object reference.

I just didn't know, how to generate a `extern "Rust" fn` in `compiler/rustc_codegen_llvm/src/allocator.rs`.

Reply:
I think the symmetric thing to do (when compared with `#[global_allocator]`) would be for the `#[alloc_error_handler]` attribute to generate `__rg_oom`, but I guess it doesn't really matter.

It is just somewhat dissatisfying that the very same problem is solved three times in slightly different ways with various amounts of code sharing (global allocator, alloc error handler, panic handler).

Reply:
> I think the symmetric thing to do (when compared with `#[global_allocator]`) would be for the `#[alloc_error_handler]` attribute to generate `__rg_oom`, but I guess it doesn't really matter.
> 
> It is just somewhat dissatisfying that the very same problem is solved three times in slightly different ways with various amounts of code sharing (global allocator, alloc error handler, panic handler).

I agree, but to implement it efficiently, I lack the expertise to redirect it via the rust ABI.